src/serac/numerics/functional/tuple_tensor_dual_functions.hpp

```diff
@@ -754,4 +754,161 @@ auto eigenvalues(const serac::tensor<T, size, size>& A)
   return output;
 }
 
+/**
+ * @brief Signum, returns sign of input
+ *
+ * @param val Input value.
+ * @return 0 when input is negative, 0 when input is 0, 1 when input is positive.
+ */
+template <typename T>
+int sgn(T val)
+{
+  // Should we implement the derivative?
+  // It should be NaN when val = 0
+  return (T(0) < val) - (val < T(0));
+}
+
+/**
+ * @brief Find indices that would sort a 3-vector
+ *
+ * @param v 3-vector to sort.
+ * @return 3-vector of indices that would sort \p v in ascending order.
+ */
+template <typename T>
+SERAC_HOST_DEVICE tensor<int, 3> argsort(const tensor<T, 3>& v)
+{
+  auto swap = [](int& first, int& second) {
+    int tmp = first;
+    first   = second;
+    second  = tmp;
+  };
+  tensor<int, 3> order{0, 1, 2};
+  if (v[0] > v[1]) swap(order[0], order[1]);
+  if (v[order[1]] > v[order[2]]) swap(order[1], order[2]);
+  if (v[order[0]] > v[order[1]]) swap(order[0], order[1]);
+  return order;
+}
+
+/** Eigendecomposition for a symmetric 3x3 matrix
+ *
+ * @param A Matrix for which the eigendecomposition will be computed. Must be
+ * symmetric, this is not checked.
+ * @return tuple with the eigenvalues in the first element, and the matrix of
+ * eigenvectors (columnwise) in the second element.
+ *
+ * @note based on "A robust algorithm for finding the eigenvalues and
+ * eigenvectors of 3x3 symmetric matrices", by Scherzinger & Dohrmann
+ */
+inline SERAC_HOST_DEVICE tuple<vec3, mat3> eig_symm(const mat3& A)
+{
+  // We know of optimizations for this routine. When this becomes the
+  // bottleneck, we can revisit. See OptimiSM for details.
+
+  tensor<double, 3>    eta{};
+  tensor<double, 3, 3> Q = DenseIdentity<3>();
+
+  auto   A_dev = dev(A);
+  double J2    = 0.5 * inner(A_dev, A_dev);
+  double J3    = det(A_dev);
+
+  if (J2 > 0.0) {
+    // angle used to find eigenvalues
+    double tmp   = (0.5 * J3) * std::pow(3.0 / J2, 1.5);
+    double alpha = std::acos(fmin(fmax(tmp, -1.0), 1.0)) / 3.0;
+
+    // consider the most distinct eigenvalue first
+    if (6.0 * alpha < M_PI) {
+      eta[0] = 2 * std::sqrt(J2 / 3.0) * std::cos(alpha);
+    } else {
+      eta[0] = 2 * std::sqrt(J2 / 3.0) * std::cos(alpha + 2.0 * M_PI / 3.0);
+    }
```

Comment:
If anyone else was surprised to see trigonometry and angles show up when solving a cubic equation, here's an animation of the underlying idea:
![cubic_roots](https://github.com/LLNL/serac/assets/61714427/72b1fc0d-b0f7-4223-b2fe-301e6e81f9cf)

The cubic roots can be expressed as x-coordinates of 3 points (120 degrees apart) on a circle with center `tr(A) / 3.0` and radius `2 * std::sqrt(J2 / 3.0)`.